plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/LazyConnectionFactory.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class LazyConnectionFactory
+        implements ConnectionFactory
+{
+    private final ConnectionFactory delegate;
+
+    public LazyConnectionFactory(ConnectionFactory delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        return new LazyConnection(() -> delegate.openConnection(session));
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        delegate.close();
+    }
+
+    private static final class LazyConnection
+            extends ForwardingConnection
+    {
+        private final SqlSupplier<Connection> connectionSupplier;
+        private volatile Connection connection;
```

Comment:
nit: volatile not needed if unless it is allowed for `Connection` to be used by multiple threads without synchronisations. I would be supprised it is allowed but I did not find obvious indication in Javadoc so not sure.

Reply:
I think JDBC requires Connection to be thread safe and there can be things like async abort or close.

However, regular `synchronized` would be as good and would yield for simpler code.
I don't know whether http://www.javaperformancetuning.com/news/qotm051.shtml is a reputable source, but they claim `synchronized` can be elided but volatile read cannot, so... maybe just keep the code simpler.